internal/controllers/watcher/controller.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to start manager: %w", err)
+	}
+
+	if err = (&WatcherReconciler{
+		Client:        mgr.GetClient(),
+		Log:           l.WithName("controllers").WithName("mcoa-watcher"),
+		Scheme:        mgr.GetScheme(),
+		addonnManager: addonManager,
+	}).SetupWithManager(mgr); err != nil {
+		return nil, fmt.Errorf("unable to create mcoa-watcher controller: %w", err)
+	}
+
+	if err = mgr.AddHealthzCheck("health", healthz.Ping); err != nil {
+		return nil, fmt.Errorf("unable to set up health check: %w", err)
+	}
+	if err = mgr.AddReadyzCheck("check", healthz.Ping); err != nil {
+		return nil, fmt.Errorf("unable to set up ready check: %w", err)
+	}
+
+	wm := WatcherManager{
+		mgr:    &mgr,
+		logger: l,
+	}
+
+	return &wm, nil
+}
+
+func (wm *WatcherManager) Start(ctx context.Context) {
+	wm.logger.Info("Starting watcher manager")
+	go func() {
+		err := (*wm.mgr).Start(ctx)
+		if err != nil {
+			wm.logger.Error(err, "there was an error while running the reconciliation watcher")
+		}
+	}()
+}
+
+// WatcherReconciler triggers reconciliation in the AddonManager when something changes in an upstream resource
+type WatcherReconciler struct {
+	client.Client
+	Log           logr.Logger
+	Scheme        *runtime.Scheme
+	addonnManager *addonmanager.AddonManager
+}
+
+var (
+	noNameErr      = errors.New("no name for reconciliation request")
+	noNamespaceErr = errors.New("no namespace for reconciliation request")
+)
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if req.Name == "" {
+		return ctrl.Result{}, noNameErr
+	}
+	if req.Namespace == "" {
+		return ctrl.Result{}, noNamespaceErr
+	}
+	(*r.addonnManager).Trigger(req.Namespace, req.Name)
+
+	r.Log.V(2).Info("reconciliation triggered", "cluster", req.Namespace)
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *WatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&addonapiv1alpha1.ManagedClusterAddOn{}, noReconcilePred).
+		Watches(&corev1.Secret{}, r.enqueueForClusterSpecificResource(), builder.OnlyMetadata).
+		Complete(r)
+}
+
+func (r *WatcherReconciler) enqueueForClusterSpecificResource() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
+		key := client.ObjectKey{Name: "multicluster-observability-addon", Namespace: obj.GetNamespace()}
+		addon := &addonapiv1alpha1.ManagedClusterAddOn{}
+		if err := r.Client.Get(ctx, key, addon); err != nil {
+			if apierrors.IsNotFound(err) {
+				return r.getSecretReconcileRequests(ctx, obj, addon)
+			}
+			r.Log.Error(err, "Error getting managedclusteraddon resources in event handler")
+			return nil
+		}
```

Comment:
Considering the fact that with #48 we change the behavior of `ClusterManagementAddon` to delegate creation and ownership of the `ManagedClusterAddon` resource per `ManagedCluster` to the `addon-manager` (see [here](https://github.com/rhobs/multicluster-observability-addon/pull/48/files#diff-636ef72739993fd52f63049ba5ff43b3bcf743e35124ac7c392405d6d9d5a99cR6)), the `IsNotFound` condition is not applying anymore. Means by default the `ManagedClusterAddon` will always be there and we need to traverse through the ManifestWorks everytime.

Reply:
Chiming in here, the `IsNotFound` condition can still happen as this enqueue function is called on `Secrets`. So if we look for a `ManagedClusterAddOn` resource in the same namespace as the `Secret` we can either have a hit if the Secret lives in the same namespace as the `ManagedClusterAddOn` or a miss if the `Secret` lives in the same namespace as the CLF/OTELCol in which case it's being used by possibly multiple `ManagedClusterAddOn` and in that case we need to look at the ManifestWorks.

Reply:
Oh IC, was not immediately visible when reading the code. But this is my bad mistake not being expert on the addon framework stuff yet